python/ray/serve/_private/replica.py

```diff
@@ -399,14 +400,67 @@ def _wrap_user_method_call(self, request_metadata: RequestMetadata):
         if user_exception is not None:
             raise user_exception from None
 
-    async def handle_request(
+    async def handle_request_with_rejection(
         self,
         pickled_request_metadata: bytes,
         *request_args,
         **request_kwargs,
+    ) -> AsyncGenerator[Any, None]:
+        """Entrypoint for all requests with strict max_concurrent_queries enforcement.
+
+        The first response from this generator is always a system message indicating
+        if the request was accepted (the replica has capacity for the request) or
+        rejected (the replica is already at max_concurrent_queries).
+
+        For non-streaming requests, there will only be one more message, the unary
+        result of the user request handler.
+
+        For streaming requests, the subsequent messages will be the results of the
+        user request handler (which must be a generator).
+        """
+        request_metadata = pickle.loads(pickled_request_metadata)
+        limit = self._deployment_config.max_concurrent_queries
+        num_ongoing_requests = self.get_num_ongoing_requests()
+        if num_ongoing_requests >= limit:
+            logger.warning(
+                f"Replica at capacity of max_concurrent_queries={limit}, "
+                f"rejecting request {request_metadata.request_id}."
+            )
+            yield pickle.dumps(
+                ReplicaQueueLengthInfo(
+                    accepted=False, num_ongoing_requests=num_ongoing_requests
+                )
+            )
+            return
+
+        yield pickle.dumps(
+            ReplicaQueueLengthInfo(
+                accepted=True,
+                # We increment num_ongoing_requests here to include this request,
+                # as it isn't included in the metrics until the handlers below.
+                num_ongoing_requests=num_ongoing_requests + 1,
+            )
+        )
```

Comment:
yielding does not give up the loop. here's a toy program you can use to verify it:

```python
import asyncio

async def background_task():
    while True:
        print("I GOT THE LOOP!!!!!!")
        await asyncio.sleep(0.1)

async def gen():
    i = 0
    while True:
        yield i
        i += 1

async def main():
    loop = asyncio.get_running_loop()
    loop.create_task(background_task())

    async for item in gen():
        if item % 10000000 == 0:
            print("gen at", item)

            # If you uncomment this line, you'll see "I GOT THE LOOP" printed:
            # await asyncio.sleep(0.1)


asyncio.run(main())
```

Reply:
Woah, this is pretty surprising. My understanding of `async for` matched [the Python docs](https://docs.python.org/3/reference/compound_stmts.html#the-async-for-statement) where `await` is implicitly called on each iteration. So I thought it would yield the event loop on each iteration.

Reply:
where does it say that `await` is implicitly called on each iteration?

Reply:
but yes I agree, it was surprising to me when I first learned it too :)

Reply:
It says it here:

<img width="815" alt="Screenshot 2024-02-05 at 3 06 28 PM" src="https://github.com/ray-project/ray/assets/92341594/05177b01-a6e2-49a6-899a-0e46cb2705e7">

Reply:
Yeah I had the same understanding when I first started working on the streaming logic :D 

This is definitely a possible "gotcha" in the future, so let me adjust the comment to make that clear.

Reply:
I've actually just refactored to avoid relying on this -- now we don't yield until entering the `with self._wrap_user_call` context manager.